modules/trusted-firmware-m/CMakeLists.txt

```diff
@@ -180,35 +205,34 @@ function(trusted_firmware_build)
     TFM_S_NS_SIGNED_BIN_FILE ${TFM_S_NS_SIGNED_BIN_FILE} # Merged TFM Secure/Nonsecure FW (signed)
     )
 
-  add_library(tfm_api
-    ${ZEPHYR_BASE}/modules/trusted-firmware-m/src/zephyr_tfm_log.c
-  )
+  zephyr_library_named(tfm_api)
 
-  target_include_directories(tfm_api
-    PRIVATE
-    ${ZEPHYR_TRUSTED_FIRMWARE_M_MODULE_DIR}/tf-m-tests/CMSIS/RTOS2/Include
-    PUBLIC
+  zephyr_sources(
+    src/zephyr_tfm_log.c
+    interface/interface.c
+    )
+  # Non-Secure interface to request system reboot
+  zephyr_sources_ifdef(CONFIG_TFM_PARTITION_PLATFORM src/reboot.c)
+  zephyr_sources_ifndef(CONFIG_TFM_PSA_TEST_NONE src/zephyr_tfm_psa_test.c)
+
+  zephyr_include_directories(
     ${ZEPHYR_TRUSTED_FIRMWARE_M_MODULE_DIR}/trusted-firmware-m/interface/include
-    INTERFACE
     ${TFM_GENERATED_INCLUDES}
   )
 
-  target_link_libraries(tfm_api
-    PUBLIC
-    zephyr_interface
-    INTERFACE
+  zephyr_link_libraries(
+    ${PSA_TEST_VAL_FILE}
+    ${PSA_TEST_PAL_FILE}
+    ${PSA_TEST_COMBINE_FILE}
+    ${PLATFORM_NS_FILE}
     ${TFM_API_NS_PATH}
     ${VENEERS_FILE}
-    $<TARGET_FILE:tfm_api>
   )
 
   add_dependencies(tfm_api tfm)
+  add_dependencies(app tfm_api)
```

Comment:
@tejlmand This feels a bit hacky, but I had to do it so that certain include files were available to the app when it was compiled. Is there a better way?

Reply:
you're right, this is not a good way.

First, `add_dependencies()` doesn't ensure that includes from tfm_api is inherited, it only ensures tfm_api and it's dependencies are built first.

In Zephyr, the right approach would be to to create a dedicated `tfm_interface` lib on which you can then specify the needed includes, and then have a corresponding Kconfig `APP_LINK_WITH_TFM` which defaults to `y`.
The reason why users should be able to deselect this, is because they might decide a different structure in their app, and thus link to another lib than app.

Something like:
```
zephyr_interface_library_named(tfm_interface)
target_include_directories(tfm_interface INTERFACE
    ${ZEPHYR_TRUSTED_FIRMWARE_M_MODULE_DIR}/trusted-firmware-m/interface/include
    ${TFM_GENERATED_INCLUDES}
)
```

and then in Kconfig, you would have:
```
config APP_LINK_WITH_TFM_INTERFACE
        bool "Link 'app' with TFM_INTERFACE"
         default y
        help
          Add TFM_INTERFACE header files to the 'app' include path. It may be
          disabled if the include paths for TFM_INTERFACE are causing aliasing
          issues for 'app'.
```

if the `tfm_api` also needs the same headers, you may link that directly using:
```
zephyr_library_link_libraries(tfm_interface)
```

Feel free to adjust the name `tfm_interface` if you have a better suggestion.

Reply:
See https://github.com/zephyrproject-rtos/zephyr/blob/facd0f57db85163432bdca7767a05e3f1d742b9f/cmake/extensions.cmake#L544-L568 for additional info.

Reply:
Sorry, that was imprecise wording on my part. I added the `add_dependencies` because the include files are _generated_ during the TFM build. The linking actually works fine.

Reply:
> @tejlmand Sorry, that was imprecise wording on my part. I added the `add_dependencies` because the include files are _generated_ during the TFM build. The linking actually works fine.

Thanks for clearing that up.
The `APP_LINK_WITH_` could be made so that it both ensures that includes are inherited (not needed in this case) and that TFM build is included first (which is what you want).

But maybe there is an even simpler way.
Have you tried to do:
```
add_dependency(zephyr_interface tfm)
```
and then remove:
```
  add_dependencies(tfm_api tfm)
  add_dependencies(app tfm_api)
```
?

Reply:
>`add_dependencies(zephyr_interface tfm)`

Thanks, that worked.

Reply:
@tejlmand Did you want other changes here as well, or is this good?